vars/jira.groovy

```diff
@@ -1,35 +1,52 @@
 #!/usr/bin/env groovy
 
+def lastSuccessfulBuild(passedBuilds, build) {
+    if ((build != null) && (build.result != 'SUCCESS')) {
+        echo "Finding last succesful build"
+        passedBuilds.add(build)
+        echo "Found a build"
+        lastSuccessfulBuild(passedBuilds, build.getPreviousBuild())
+    }
+}
+
 @NonCPS
-def getChangeString() {
-    MAX_MSG_LEN = 100
+def getChangeString(passedBuilds) {
     def changeString = ""
-    echo "Gathering SCM changes"
-    def changeLogSets = currentBuild.changeSets
-    for (int i = 0; i < changeLogSets.size(); i++) {
-        def entries = changeLogSets[i].items
-        for (int j = 0; j < entries.length; j++) {
-            def entry = entries[j]
-            truncated_msg = entry.msg.take(MAX_MSG_LEN)
-            changeString += " - ${truncated_msg} [${entry.author}]\n"
+    echo "Gathering Changes Since Last Successful Build"
+    for (int x = 0; x < passedBuilds.size(); x++) {
+        def currentBuild = passedBuilds[x];
+        def buildNumber = currentBuild.number
+        echo "Changes for Build ${buildNumber}"
+        def changeLogSets = currentBuild.rawBuild.changeSets
+        for (int i = 0; i < changeLogSets.size(); i++) {
+            def entries = changeLogSets[i].items
+            for (int j = 0; j < entries.length; j++) {
+                def entry = entries[j]
+                changeString += "* ${entry.msg} by ${entry.author} \n"
+            }
         }
     }
     if (!changeString) {
         changeString = " - No new changes"
     }
-    return changeString
+    echo changeString
+    return changeString;
 }
 
 @NonCPS
```

Comment:
why this is NonCPS?

Reply:
As described here: https://github.com/jenkinsci/workflow-cps-plugin/blob/master/README.md. NonCPS methods are best used to perform some calculation before passing a summary back to the main script. Since we only need a small portion of the changeset, ie the commit message and the author and that's what the method returns. It fulfills this criteria hence making it a NonCPS.